python2: do not resolve requirements if no python targets

tdyas opened the pull request:
If Python targets depend on resource targets, the resolve_requirements task will be invoked even if the current build does not have any Python targets in the targets to be processed. The resolve_requirements task should not do any work in that case.

Fixes https://github.com/pantsbuild/pants/issues/4688

benjyw commented:
Addresses https://github.com/pantsbuild/pants/issues/4688.

benjyw commented:
@tdyas Thanks for the fix! However, looking at this more deeply, I've realized that this is not the right fix after all. We do actually want to create a requirements pex in all cases, even when there are no targets, so that downstream tasks don't have to do a bunch of None checking.

The bug is something else: the task sees a directory that it thinks it previously created a pex in, but when it tries to materialize the pex, it errors because that directory doesn't actually contain a pex. 
 
The only cause I can think of is that we don't create the directory and its contents atomically.  But I don't see how that can happen, since we create it using `safe_concurrent_creation`.

Anyway, if you clean and re-run, does it reproduce?

tdyas commented:
It may have to do with selection of the Python interpreter. I changed my python-setup:interpreter_constraints from `python==2.7.13` to `cpython==2.7.13` and Pants finally used the interpreter in the virtualenv in which I am running the pants invocation. (It had been finding the macOS system Python (2.7.10) for some reason.)

To answer your question, I cleaned again and put the constraint back to `python==2.7.13` and now I get: `FAILURE: Unable to detect a suitable interpreter for compatibilities:  (Conflicting targets: )`  It is no longer using the system Python.

Inconsistent state in my .pants.d?

benjyw commented:
Closing as I've figured out the issue - `safe_concurrent_creation` renames in the `finally` block, meaning it'll rename even if there was an error while creating the content of the dir...
